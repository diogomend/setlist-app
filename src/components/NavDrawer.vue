<template>
  <v-navigation-drawer
    v-model="drawer"
    fixed
    :expand-on-hover="!isMobile"
    :temporary="isMobile"
    :mini-variant="!isMobile"
    :permanent="!isMobile"
  >
    <v-list nav>
      <v-list-item two-line :class="!isMobile && 'px-0'">
        <v-list-item-avatar>
          <img src="@/assets/avatar.jpeg" />
        </v-list-item-avatar>

        <v-list-item-content>
          <v-list-item-title>Diogo Mendon√ßa</v-list-item-title>
          <v-list-item-subtitle>My Setlists</v-list-item-subtitle>
        </v-list-item-content>
      </v-list-item>

      <v-divider></v-divider>
      <v-list-item-group v-model="group" active-class="text--accent-4">
        <v-list-item
          v-for="(page, index) in pages"
          :key="index"
          :to="page.link"
        >
          <v-list-item-icon>
            <v-icon>{{ page.icon }}</v-icon>
          </v-list-item-icon>
          <v-list-item-title>{{ page.text }}</v-list-item-title>
        </v-list-item>
      </v-list-item-group>
    </v-list>
  </v-navigation-drawer>
</template>

<script>
import { mapActions, mapGetters } from "vuex";
import mobileMixin from "@/mixins/mobileMixin";
export default {
  name: "NavDrawer",
  mixins: [mobileMixin],
  data() {
    return {
      group: true,
      pages: [
        {
          icon: "home",
          text: "Home",
          link: "/"
        },
        {
          icon: "timeline",
          text: "Shows",
          link: "/shows"
        },
        {
          icon: "person",
          text: "Artists",
          link: "/artists"
        }
      ]
    };
  },
  computed: {
    drawer: {
      get() {
        return this.$store.state.appState.drawer;
      },
      set(value) {
        this.toggleDrawer(value);
      }
    },
    ...mapGetters({
      isOpened: "setlist/isLoading"
    })
  },
  methods: {
    ...mapActions("appState", ["toggleDrawer"])
  }
};
</script>

<style scoped>
.v-navigation-drawer--absolute {
  z-index: 10;
}
</style>
