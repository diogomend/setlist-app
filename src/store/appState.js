const actions = {
  toggleDrawer: ({ commit }, toggle) => {
    commit("TOGGLE_DRAWER", toggle);
  },
  toggleLoading: ({ commit }, toggle) => {
    commit("TOGGLE_LOADING", toggle);
  }
};

const mutations = {
  TOGGLE_DRAWER(state, bool) {
    state.drawer = bool;
  },
  TOGGLE_LOADING(state, isLoading) {
    state.isLoading = isLoading;
  }
};

const getters = {
  getDrawer: state => state.drawer,
  getLoading: state => state.isLoading
};

const state = () => ({
  drawer: false,
  isLoading: true
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
