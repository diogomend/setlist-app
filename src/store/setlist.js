import setlistService from "@/service/setlist";

const actions = {
  fetchSetlists: async ({ commit, dispatch }) => {
    try {
      const { data } = await setlistService.getSetlists();
      commit("UPDATE_SETLISTS", data);
      commit("UPDATE_ARTISTS", data);
      //return Promise.resolve(data);
      return true;
    } catch (err) {
      //return Promise.reject('Could not load setlist information');
      return false;
    } finally {
      dispatch("appState/toggleLoading", false, { root: true })
    }
  }
};

const mutations = {
  UPDATE_SETLISTS(state, data) {
    state.setlists = data;
  },
  UPDATE_ARTISTS(state, data) {
    state.artists = getArtists(data);
  }
};

const getters = {
  getTotalConcerts: state => state.setlists.length,
  getShow: state => id => {
    return Object.values(state.setlists).find(element => element.id === id);
  },
  getArtist: state => mbid => {
    const data = Object.values(state.setlists).find(
      element => element.artist.mbid === mbid
    );
    if (data && Object.prototype.hasOwnProperty.call(data, "artist")) {
      return data.artist;
    }

    return null;
  },
  getArtists: state => {
    const artists = getArtists(state.setlists);

    return Object.values(artists).sort((a, b) => (a.name > b.name ? 1 : -1));
  },
  getArtistShows: state => mbid => {
    return state.setlists.reduce((ac, setlist) => {
      if (setlist.artist.mbid === mbid) {
        ac.push(setlist);
      }

      return ac;
    }, []);
  },
  getTopArtists: state => {
    const artists = Object.values(state.artists).sort((a, b) =>
      a.qty < b.qty ? 1 : -1
    );
    return artists.slice(0, 6);
  },
  getTotalArtists: state => Object.keys(state.artists).length,
  getYearlyConcerts: state => {
    return state.setlists.reduce((ac, setlist) => {
      const dateSplit = setlist.eventDate.split("-");
      const year = parseInt(dateSplit[2]);

      if (Object.prototype.hasOwnProperty.call(ac, year)) {
        const merge = [...ac[year], setlist];
        return { ...ac, [year]: merge };
      }

      return { ...ac, [year]: [setlist] };
    }, {});
  },
  getTotalCountries: state => {
    const countries = state.setlists.reduce((ac, setlist) => {
      const country = setlist.venue.city.country.code;
      const qty = Object.prototype.hasOwnProperty.call(ac, country)
        ? ac[country] + 1
        : 1;

      return { ...ac, [country]: qty };
    }, {});

    return Object.keys(countries).length;
  }
};

const state = () => ({
  setlists: [],
  isLoading: true
});

const getArtists = setlists => {
  return setlists.reduce((ac, setlist) => {
    setlist.artist.qty = Object.prototype.hasOwnProperty.call(
      ac,
      setlist.artist.mbid
    )
      ? ac[setlist.artist.mbid].qty + 1
      : 1;

    return { ...ac, [setlist.artist.mbid]: setlist.artist };
  }, {});
};

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
