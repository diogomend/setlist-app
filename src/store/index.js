import Vue from "vue";
import Vuex from "vuex";
import setlist from "@/store/setlist";
import appState from "@/store/appState";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    setlist,
    appState
  }
};

export default new Vuex.Store(storeOptions);
