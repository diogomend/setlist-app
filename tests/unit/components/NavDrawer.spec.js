import { shallowMount } from "@vue/test-utils";
import Vue from "vue";
import { storeOptions } from "@/store/index";
import NavDrawer from "@/components/NavDrawer";
import Vuex from "vuex";
import Vuetify from "vuetify";

Vue.use(Vuex);
Vue.use(Vuetify);
describe("NavDrawer.vue", () => {
  let store;
  let localVue;

  beforeEach(() => {
    jest.clearAllMocks();
    store = new Vuex.Store(storeOptions);
  });

  it("is a Vue instance", () => {
    const wrapper = shallowMount(NavDrawer, {
      localVue,
      store
    });

    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });

  it("set drawer should call toggleDrawer with new value", () => {
    const toggleDrawer = jest.fn();
    const wrapper = shallowMount(NavDrawer, {
      localVue,
      store
    });

    wrapper.setMethods({ toggleDrawer });
    expect(wrapper.vm.drawer).toBeFalsy();
    wrapper.vm.drawer = true;
    expect(toggleDrawer).toHaveBeenCalledWith(true);
  });
});
