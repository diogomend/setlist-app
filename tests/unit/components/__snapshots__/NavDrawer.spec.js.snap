// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavDrawer.vue is a Vue instance 1`] = `
<v-navigation-drawer-stub fixed="true" expandonhover="true" height="100%" minivariant="true" minivariantwidth="56" mobilebreakpoint="1264" permanent="true" src="" tag="aside" width="256">
  <v-list-stub tag="div" tile="true" nav="true">
    <v-list-item-stub activeclass="" tag="div" twoline="true" class="px-0">
      <v-list-item-avatar-stub size="40"><img src="@/assets/avatar.jpeg"></v-list-item-avatar-stub>
      <v-list-item-content-stub>
        <v-list-item-title-stub>Diogo Mendon√ßa</v-list-item-title-stub>
        <v-list-item-subtitle-stub>My Setlists</v-list-item-subtitle-stub>
      </v-list-item-content-stub>
    </v-list-item-stub>
    <v-divider-stub></v-divider-stub>
    <v-list-item-group-stub value="true" activeclass="text--accent-4">
      <v-list-item-stub activeclass="" to="/" tag="div">
        <v-list-item-icon-stub>
          <v-icon-stub>home</v-icon-stub>
        </v-list-item-icon-stub>
        <v-list-item-title-stub>Home</v-list-item-title-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" to="/shows" tag="div">
        <v-list-item-icon-stub>
          <v-icon-stub>timeline</v-icon-stub>
        </v-list-item-icon-stub>
        <v-list-item-title-stub>Shows</v-list-item-title-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" to="/artists" tag="div">
        <v-list-item-icon-stub>
          <v-icon-stub>person</v-icon-stub>
        </v-list-item-icon-stub>
        <v-list-item-title-stub>Artists</v-list-item-title-stub>
      </v-list-item-stub>
    </v-list-item-group-stub>
  </v-list-stub>
</v-navigation-drawer-stub>
`;
